Lib/test/test_concurrent_futures/test_process_pool.py

```diff
@@ -218,6 +237,106 @@ def mock_start_new_thread(func, *args, **kwargs):
+            worker_process = list(executor._processes.values())[0]
+
+            Mock = unittest.mock.Mock
+            worker_process.terminate = Mock(wraps=worker_process.terminate)
+            worker_process.kill = Mock(wraps=worker_process.kill)
+
+            getattr(executor, function_name)()
+            worker_process.join()
+
+            if function_name == TERMINATE_WORKERS:
+                worker_process.terminate.assert_called()
+            elif function_name == KILL_WORKERS:
+                worker_process.kill.assert_called()
+            else:
+                self.fail(f"Unknown operation: {function_name}")
+
+            self.assertRaises(queue.Empty, q.get, timeout=0.01)
+
+    @parameterize(*FORCE_SHUTDOWN_PARAMS)
+    def test_force_shutdown_workers_dead_workers(self, function_name):
+        with self.executor_type(max_workers=1) as executor:
+            future = executor.submit(os._exit, 1)
+            self.assertRaises(BrokenProcessPool, future.result)
+
+            # even though the pool is broken, this shouldn't raise
+            getattr(executor, function_name)()
+
+    @parameterize(*FORCE_SHUTDOWN_PARAMS)
+    def test_force_shutdown_workers_not_started_yet(self, function_name):
+        ctx = self.get_context()
+        with unittest.mock.patch.object(ctx, 'Process') as mock_process:
+            with self.executor_type(max_workers=1, mp_context=ctx) as executor:
+                # The worker has not been started yet, terminate/kill_workers
+                # should basically no-op
+                getattr(executor, function_name)()
+
+            mock_process.return_value.kill.assert_not_called()
+            mock_process.return_value.terminate.assert_not_called()
+
+    @parameterize(*FORCE_SHUTDOWN_PARAMS)
+    def test_force_shutdown_workers_stops_pool(self, function_name):
+        with self.executor_type(max_workers=1) as executor:
+            task = executor.submit(time.sleep, 0)
+            self.assertIsNone(task.result())
+
+            worker_process = list(executor._processes.values())[0]
+            getattr(executor, function_name)()
+
+            self.assertRaises(RuntimeError, executor.submit, time.sleep, 0)
+
+            # A signal sent, is not a signal reacted to.
+            # So wait a moment here for the process to die.
+            # If we don't, every once in a while we may get an ENV CHANGE
+            # error since the process would be alive immediately after the
+            # test run.. and die a moment later.
+            worker_process.join(support.SHORT_TIMEOUT)
+
+            # Oddly enough, even though join completes, sometimes it takes a
+            # moment for the process to actually be marked as dead.
+            # ...  that seems a bit buggy.
+            # We need it dead before ending the test to ensure it doesn't
+            # get marked as an ENV CHANGE due to living child process.
+            while worker_process.is_alive():
+                time.sleep(0.1)
```

Comment:
I think we should use something like:

```python
for _ in support.sleeping_retry(support.SHORT_TIMEOUT):
    if not worker_process.is_alive():
        break
```

That has a few nice properties:

* It will eventually raise an exception instead of looping forever, if for some reason, `worker_process.is_alive()` isn't False after a reasonable amount of time.
* The `time.sleep()` starts out very small and grows